node/impl/full/actor_events.go

```diff
@@ -195,54 +237,41 @@ func (a *ActorEventHandler) SubscribeActorEvents(ctx context.Context, evtFilter
 			// tell the caller we're done
 			close(out)
 			fm.ClearSubChannel()
-			if err := a.EventFilterManager.Remove(ctx, fm.ID()); err != nil {
+			if err := a.eventFilterManager.Remove(ctx, fm.ID()); err != nil {
 				log.Warnf("failed to remove filter: %s", err)
 			}
 		}()
 
+		// When we start sending real-time events, we want to make sure that we don't fall behind more
+		// than one epoch's worth of events (approximately). Capture this value now, before we send
+		// historical events to allow for a little bit of slack in the historical event sending.
+		minBacklogHeight := a.chain.GetHeaviestTipSet().Height() - 1
+
 		// Handle any historical events that our filter may have picked up -----------------------------
 
-		evs, minEpoch, maxEpoch := getCollected(ctx, fm)
+		evs := getCollected(ctx, fm)
 		if len(evs) > 0 {
-			// must be able to send events at least twice as fast as they were generated
-			epochRange := maxEpoch - minEpoch
-			if epochRange <= 0 {
-				epochRange = 1
-			}
-			eventsPerEpoch := float64(len(evs)) / float64(epochRange)
-			eventsPerSecond := 2 * eventsPerEpoch / float64(build.BlockDelaySecs)
-			// a minimum rate of 1 event per second if we don't have many events
-			if eventsPerSecond < 1 {
-				eventsPerSecond = 1
-			}
-
-			// send events from evs to the out channel and ensure we don't do it slower than eventsPerMs
-			ticker := time.NewTicker(time.Second)
-			defer ticker.Stop()
-
-			const maxSlowTicks = 3 // slightly forgiving, allow 3 slow ticks (seconds) before giving up
-			slowTicks := 0
-			sentEvents := 0.0
-
+			// ensure we get all events out on the channel within one block's time (30s on mainnet)
+			timer := a.clock.Timer(a.blockDelay)
 			for _, ev := range evs {
 				select {
 				case out <- ev:
-					sentEvents++
-				case <-ticker.C:
-					if sentEvents < eventsPerSecond {
-						slowTicks++
-						if slowTicks >= maxSlowTicks {
-							log.Errorf("closing event subscription due to slow event sending rate")
-							return
-						}
-					} else {
-						slowTicks = 0
-					}
-					sentEvents = 0
+				case <-timer.C:
+					log.Errorf("closing event subscription due to slow event sending rate")
+					timer.Stop()
+					return
 				case <-ctx.Done():
+					timer.Stop()
 					return
 				}
 			}
+			timer.Stop()
+		}
+
+		// for the case where we have a MaxHeight set, we don't get a signal from the filter when we
+		// reach that height, so we need to check it ourselves, do it now but also in the loop
+		if params.MaxHeight > 0 && a.chain.GetHeaviestTipSet().Height() > params.MaxHeight {
```

Comment:
This point is the same for the `nextBacklogHeightUpdate` comment below: in the case of a restrictive filter, we may not get an event to use for our check for a long time, or ever, so we can't rely on looking at new events to check for height. e.g. the itest in here won't terminate if we don't have these checks because it does a subscribe with a height termination condition and no further activity. This is a reasonable case when you consider how specific a filter can be.

Reply:
Hm. You're right. Can we check _before_ we start the filter? My concern here and above is that we:

1. Install the filter.
2. _Then_ we get the tipset height.

The tipset could advance in the meantime and we could end up missing some events.

I mean, we could miss events _anyways_ due to reorgs, but that's less likely. So we're probably good (after filing an issue) to land this as long as we check the tipset height first.